src/metabase/api/cloud_migration.clj

```diff
@@ -0,0 +1,249 @@
+(ns metabase.api.cloud-migration
+  "/api/cloud-migration endpoints.
+  Only one migration should be happening at any given time.
+  But if something weird happens with concurrency, /cancel will
+  cancel all of them. "
+  (:require
+   [cheshire.core :as json]
+   [clj-http.client :as http]
+   [clojure.java.io :as io]
+   [clojure.set :as set]
+   [compojure.core :refer [GET POST PUT]]
+   [metabase.api.common :as api]
+   [metabase.cmd.dump-to-h2 :as dump-to-h2]
+   [metabase.config :as config]
+   [metabase.models.cloud-migration :refer [CloudMigration]]
+   [metabase.models.setting :refer [defsetting]]
+   [metabase.models.setting.cache :as setting.cache]
+   [metabase.util.i18n :refer [deferred-tru]]
+   [metabase.util.log :as log]
+   [methodical.core :as methodical]
+   [toucan2.core :as t2]
+   [toucan2.pipeline :as t2.pipeline])
+  (:import
+   [java.io File InputStream]))
+
+(set! *warn-on-reflection* true)
+
+(defsetting metabase-store-migration-url
+  (deferred-tru "Store URL for migrations. Internal test use only.")
+  :visibility :internal
+  :default    "https://store-api.metabase.com/api/v2/migration"
+  :doc        false
+  :export?    false)
+
+;; Read-Only mode
+
+(defsetting read-only-mode
+  (deferred-tru
+    (str "Boolean indicating whether a Metabase's is in read-only mode with regards to its app db. "
+         "Will take up to 1m to propagate to other Metabase instances in a cluster."
+         "Audit tables are excluded from read-only-mode mode."))
+  :type       :boolean
+  :visibility :admin
+  :default    false
+  :doc        false
+  :export?    false)
+
+(def ^:private read-only-mode-exceptions
+  #{;; Migrations need to update their own state
+    :model/CloudMigration
+    :model/Setting
+
+    ;; Users need to login, make queries, and we need need to audit them.
+    :model/Session
+    :model/User
+    :model/LoginHistory
+    :model/AuditLog
+    :model/QueryExecution
+    :model/ViewLog
+
+    ;; TODO: Came up in tests, check with core folks
+    :setting
+    :permissions_group_membership})
```

Comment:
Not sure why `permissions_group_membership` would need to be writable. Why were tests failing?

`setting` is tricky, because most standard settings probably shouldn't be writable during migration either. Like run-of-the-mill admin settings. If someone updates those while the migration is ongoing, the setting would be lost, right? 

I guess that's also true of the `core_user` table—we'd presumably allow someone to change their password in read-only mode. That might be an issue.

Are we OK from the product POV for the "read-only" mode to be this coarsely grained, and possibly lose data in these cases?

Reply:
Tested `permissions_group_membership` and `core_user` again, and didn't need it anymore, so removed it. I was getting errors from it at some point, not sure what changed meanwhile.